operator-common/src/main/java/io/strimzi/operator/common/AbstractOperator.java

```diff
@@ -139,6 +142,7 @@ public AbstractOperator(Vertx vertx, String kind, O resourceOperator, MetricsPro
                 metricTags);
 
         resourcesStateCounter = new ConcurrentHashMap<>();
+        reconciliations = new ConcurrentHashMap<>();
```

Comment:
It seems to me that the map (internal to vertx) used for the locks is exactly what we need to provide the necessary semantics, and having two maps, as here, opens the door to a race (though I think it's unlikely to be a problem in practice). I'm not sure how easy it would be to implement our own replacement for the locking functionality that supported cancellation.

Reply:
Are you thinking something like `vertx.sharedData().getLocalMap("reconciliations")` ? I may not be following you completely.

Reply:
The first time I looked at this PR I was wondering whether we really needed a map of _all_ the reconciliations, or whether we were just interested in the cancelled ones. I think the original comment was motivated by this. Now I look again, I don't see the original problem. 

That the cancellation doesn't necessarily affect just the running reconciliation, but also any that are queued up waiting on the lock makes me wonder whether anything guarantees that at the time we process a DELETED event (which will cause the cancellation) that the queue cannot have a recreation ADDED event. This would be guaranteed by the assumption that fabric8 dispatches all watch events via a single thread. But off the top of my head I'm not certain whether that assumption is true. In any case I think this would be super-rare in practice and would be fixed by the next periodic reconciliation.